src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -293,22 +293,33 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
             //  6. 'pipe.Writer' is completed
             //  7. 'pipe.Reader' is completed
             //  8. OperationCanceledException is thrown back to the caller
+            //
+            // Since StreamJsonRpc will automatically complete writers when an error occurs (and cancellation is
+            // treated as an error), we create a buffered reader to isolate callback actions from this
+            // implementation.
 
             // Create a separate cancellation token for the reader, which we keep open until after the call to invoke
             // completes. If we close the reader before cancellation is processed by the remote call, it might block
             // (deadlock) while writing to a stream which is no longer processing data.
             using var readerCancellationSource = new CancellationTokenSource();
 
-            var pipe = new Pipe();
+            var writeBufferPipe = new Pipe();
+            var readBufferPipe = new Pipe();
 
             // Create new tasks that both start executing, rather than invoking the delegates directly
             // to make sure both invocation and reader start executing and transfering data.
 
             var writerTask = Task.Run(async () =>
             {
+                var copyTask = writeBufferPipe.Reader.CopyToAsync(readBufferPipe.Writer, cancellationToken);
+
                 try
                 {
-                    await invocation(service, pipe.Writer, cancellationToken).ConfigureAwait(false);
+                    await invocation(service, writeBufferPipe.Writer, cancellationToken).ConfigureAwait(false);
+
+                    // Complete the copy and mark the operation complete
+                    await copyTask.NoThrowAwaitable(captureContext: false);
```

Comment:
If `copyTask` faults, then you may never reach this line because `invocation` may be waiting to read more bytes but no bytes are coming and you haven't completed the writer, so it just keeps waiting.
For that, you may want a clever aggregating continuation Task that will complete early if *any* of the input tasks fault. I have done something like that in a test utility method I can dig up if you'd like.

Reply:
➡️ Now rewritten